fragment.go

```diff
@@ -827,41 +880,64 @@ func (f *fragment) importSetValue(columnID uint64, bitDepth uint, value uint64,
 		}
 	}
 
+	// Mark sign bit.
+	if p, err := f.pos(uint64(bsiSignBit), columnID); err != nil {
+		return changed, errors.Wrap(err, "getting sign pos")
+	} else if value >= 0 || clear {
+		if c, err := f.storage.Remove(p); err != nil {
+			return changed, errors.Wrap(err, "removing sign from storage")
+		} else if c {
+			changed = true
+		}
+	} else {
+		if c, err := f.storage.Add(p); err != nil {
+			return changed, errors.Wrap(err, "adding sign to storage")
+		} else if c {
+			changed = true
+		}
+	}
+
 	return changed, nil
 }
 
 // sum returns the sum of a given bsiGroup as well as the number of columns involved.
 // A bitmap can be passed in to optionally filter the computed columns.
-func (f *fragment) sum(filter *Row, bitDepth uint) (sum, count uint64, err error) {
+func (f *fragment) sum(filter *Row, bitDepth uint) (sum int64, count uint64, err error) {
 	// Compute count based on the existence row.
-	consider := f.row(uint64(bitDepth))
+	consider := f.row(bsiExistsBit)
 	if filter != nil {
 		consider = consider.Intersect(filter)
 	}
 	count = consider.Count()
 
+	// Determine positive & negative sets.
+	nrow := f.row(bsiSignBit)
+	prow := consider.Difference(nrow)
+
 	// Compute the sum based on the bit count of each row multiplied by the
 	// place value of each row. For example, 10 bits in the 1's place plus
 	// 4 bits in the 2's place plus 3 bits in the 4's place equals a total
 	// sum of 30:
 	//
 	//   10*(2^0) + 4*(2^1) + 3*(2^2) = 30
 	//
-	var cnt uint64
+	// Execute once for positive numbers and once for negative. Subtract the
+	// negative sum from the positive sum.
+	var psum, nsum int64
 	for i := uint(0); i < bitDepth; i++ {
-		row := f.row(uint64(i))
-		cnt = row.intersectionCount(consider)
-		sum += (1 << i) * cnt
+		row := f.row(uint64(bsiOffsetBit + i))
+
+		psum += int64((1 << i) * row.intersectionCount(prow))
```

Comment:
this is sort of brilliant.

a thing has occurred to me, though: for large BSI ranges, for large column counts, it seems to me that there's a risk of overflow.

should we perhaps be squashing these? both in terms of filtering out intersection counts before doing this multiply, and possibly later in comparing psum and nsum; if they're both greater than zero, we can reduce them both by the lower of the two values.